src/mpid/ch4/src/ch4_recv.h

```diff
@@ -57,14 +57,19 @@ MPL_STATIC_INLINE_PREFIX int MPID_Recv(void *buf,
             MPIR_ERR_POP(mpi_errno);
         }
 
-        mpi_errno = MPIDI_NM_mpi_irecv(buf, count, datatype, rank, tag, comm, context_offset, av,
-                                       &(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request)));
+        MPIDI_CH4I_REQUEST(*request, is_local) = 1;
 
-        if (mpi_errno != MPI_SUCCESS) {
-            MPIR_ERR_POP(mpi_errno);
-        }
-        /* cancel the shm request if netmod/am handles the request from unexpected queue. */
-        else if (*request) {
+        if (!MPIR_Request_is_complete(*request) && !MPIDI_CH4U_REQUEST_IN_PROGRESS(*request)) {
+            mpi_errno = MPIDI_NM_mpi_irecv(buf, count, datatype, rank, tag, comm, context_offset,
+                                           av, &(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request)));
+
+            if (mpi_errno != MPI_SUCCESS) {
+                MPIR_ERR_POP(mpi_errno);
+            }
+
+            MPIDI_CH4I_REQUEST(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request), is_local) = 0;
+
+            /* cancel the shm request if netmod/am handles the request from unexpected queue. */
             if (MPIR_Request_is_complete(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request))) {
```

Comment:
There are in progress requests in the active message code (this fix was not just for shared memory but for all active message code). Look in `mpidig_recv.h`.

Reply:
There isn't anything in the netmod code (or the shared memory code for that matter) independently. It's all in the active message code.

Reply:
It doesn't work the other way around because we call the shared memory code first every time (because it's software matching so cancellation is easier).